File: archive/internal/auth/principal_mapping.go
package auth

import (
	"fmt"
	"regexp"
	"strings"
)

// PrincipalMapper extracts the principal name from an X.509 subject
// DN (or, in the future, SAN URI / SAN DNS) per Apache Kafka's
// `ssl.principal.mapping.rules` config (gh #43, KIP-371). Each rule
// is one of:
//
//   - "DEFAULT"     — return the subject's CommonName unchanged.
//   - "RULE:<regex>/<replacement>[/L|/U]" — match the full DN against
//     <regex>; on success, format the principal as <replacement>
//     (with $1, $2, ... back-references). Optional /L lowercases,
//     /U uppercases the result. First matching rule wins; if none
//     match, the next rule is tried in declaration order.
//
// Skafka mirrors Apache's first-match-wins, leftmost-rule precedence.
// If the rule list is empty (or no rule matches), the principal name
// defaults to the CN (pre-gh #43 behaviour, preserved as a back-stop
// so an empty config keeps clusters working unchanged).
type PrincipalMapper struct {
	rules []principalRule
}

type principalRule struct {
	regex       *regexp.Regexp
	replacement string
	caseFlag    byte // 'L', 'U', or 0
	isDefault   bool
}

// NewPrincipalMapper compiles a rule string in Apache's format.
// Rules are separated by commas. Empty input returns a mapper whose
// Apply() always returns the CN unchanged.
//
// Example: "RULE:^CN=(.*?),.*$/$1/L,DEFAULT"
//
// On parse error (invalid regex, malformed RULE syntax), returns the
// error so callers can surface it at chart-config-validate time
// rather than silently falling back to "use CN".
func NewPrincipalMapper(spec string) (*PrincipalMapper, error) {
	m := &PrincipalMapper{}
	spec = strings.TrimSpace(spec)
	if spec == "" {
		return m, nil
	}
	for _, raw := range splitRules(spec) {
		r, err := parsePrincipalRule(raw)
		if err != nil {
			return nil, fmt.Errorf("rule %q: %w", raw, err)
		}
		m.rules = append(m.rules, r)
	}
	return m, nil
}

// Apply returns the mapped principal name for an X.509 subject. The
// subject input is the full RFC 2253 DN string (e.g.
// "CN=alice,OU=Engineering,O=Acme"). cn is the convenience-extracted
// CommonName the caller already has — used as the fall-back when no
// rule matches.
func (m *PrincipalMapper) Apply(subject, cn string) string {
	if m == nil {
		return cn
	}
	for _, r := range m.rules {
		if r.isDefault {
			return cn
		}
		match := r.regex.FindStringSubmatchIndex(subject)
		if match == nil {
			continue
		}
		out := string(r.regex.ExpandString(nil, r.replacement, subject, match))
		switch r.caseFlag {
		case 'L':
			out = strings.ToLower(out)
		case 'U':
			out = strings.ToUpper(out)
		}
		return out
	}
	return cn
}

// splitRules splits the rule spec at rule boundaries. Rules are
// separated by commas, but commas can appear inside a RULE:'s regex
// body (Subject DNs use them as a separator). We split only at
// commas immediately followed by "RULE:" or "DEFAULT" so the regex
// body's internal commas stay attached to the previous rule.
//
// Apache uses the same boundary heuristic — its `SslPrincipalMapper`
// parses with a "rule starts with RULE:" prefix scan.
func splitRules(spec string) []string {
	var out []string
	i := 0
	for i < len(spec) {
		// Find the start of the next rule boundary: a comma followed
		// (after optional whitespace) by RULE: or DEFAULT.
		j := i
		for j < len(spec) {
			if spec[j] == ',' {
				k := j + 1
				for k < len(spec) && (spec[k] == ' ' || spec[k] == '\t') {
					k++
				}
				if strings.HasPrefix(spec[k:], "RULE:") || strings.HasPrefix(spec[k:], "DEFAULT") {
					break
				}
			}
			j++
		}
		out = append(out, strings.TrimSpace(spec[i:j]))
		if j < len(spec) {
			j++ // consume the comma
		}
		i = j
	}
	return out
}

// parsePrincipalRule parses one RULE: or DEFAULT entry. Apache's
// canonical format is:
//
//	RULE:<regex>/<replacement>[/L|/U]
//
// where the trailing /L|/U flag is optional and there is always a
// terminating "/" after the replacement (even when no flag follows).
// Equivalently: the rule body always ends with "/" or "/L" or "/U".
// Regexes containing literal slashes are not supported — Subject DNs
// never carry them; this matches Apache's behaviour.
func parsePrincipalRule(raw string) (principalRule, error) {
	if raw == "DEFAULT" {
		return principalRule{isDefault: true}, nil
	}
	if !strings.HasPrefix(raw, "RULE:") {
		return principalRule{}, fmt.Errorf("must start with RULE: or be DEFAULT")
	}
	body := raw[len("RULE:"):]

	// Apache's body always ends with `/`, `/L`, or `/U`. Trim
	// whichever suffix is present so the remaining `body` is the
	// raw `<regex>/<replacement>` pair.
	caseFlag := byte(0)
	switch {
	case strings.HasSuffix(body, "/L"):
		caseFlag = 'L'
		body = body[:len(body)-2]
	case strings.HasSuffix(body, "/U"):
		caseFlag = 'U'
		body = body[:len(body)-2]
	case strings.HasSuffix(body, "/"):
		body = body[:len(body)-1]
	default:
		return principalRule{}, fmt.Errorf("rule must end with '/', '/L', or '/U'")
	}

	// Split the remainder at the FIRST "/". DNs don't carry slashes,
	// so this is unambiguous in practice (matches Apache's behaviour).
	slash := strings.IndexByte(body, '/')
	if slash < 0 {
		return principalRule{}, fmt.Errorf("missing '/' separator between regex and replacement")
	}
	pattern := body[:slash]
	replacement := body[slash+1:]
	re, err := regexp.Compile(pattern)
	if err != nil {
		return principalRule{}, fmt.Errorf("compile regex: %w", err)
	}
	return principalRule{
		regex:       re,
		replacement: replacement,
		caseFlag:    caseFlag,
	}, nil
}

File: crates/sk-auth/src/principal_mapping.rs
//! `ssl.principal.mapping.rules` (gh #43, KIP-371).
//!
//! Port of `archive/internal/auth/principal_mapping.go`. Each rule is
//! one of:
//!
//! - `"DEFAULT"` — return the subject's CommonName unchanged.
//! - `"RULE:<regex>/<replacement>[/L|/U]"` — match the full DN
//!   against `<regex>`; on success, format the principal as
//!   `<replacement>` (with `$1`, `$2`, ... back-references). Optional
//!   `/L` lowercases, `/U` uppercases the result. First matching
//!   rule wins; if none match, fall through to the CN.
//!
//! Empty spec returns a mapper whose [`PrincipalMapper::apply`]
//! always returns the CN unchanged — preserves the pre-gh #43
//! behaviour so an empty config keeps clusters working.

use regex::Regex;

use crate::errors::AuthError;

#[derive(Debug, Default)]
pub struct PrincipalMapper {
    rules: Vec<Rule>,
}

#[derive(Debug)]
enum Rule {
    Default,
    Replace {
        regex: Regex,
        replacement: String,
        case_flag: CaseFlag,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum CaseFlag {
    None,
    Lower,
    Upper,
}

impl PrincipalMapper {
    /// Compile a rule string in Apache's format. Comma-separated;
    /// empty input returns a mapper whose `apply` always returns the
    /// CN unchanged.
    pub fn parse(spec: &str) -> Result<Self, AuthError> {
        let spec = spec.trim();
        if spec.is_empty() {
            return Ok(Self::default());
        }
        let mut rules = Vec::new();
        for raw in split_rules(spec) {
            rules.push(parse_one(&raw)?);
        }
        Ok(Self { rules })
    }

    /// Return the mapped principal name for an X.509 subject. `cn` is
    /// the convenience-extracted CommonName the caller already has —
    /// used as the fall-back when no rule matches and on `DEFAULT`.
    pub fn apply(&self, subject_dn: &str, cn: &str) -> String {
        for rule in &self.rules {
            match rule {
                Rule::Default => return cn.to_owned(),
                Rule::Replace {
                    regex,
                    replacement,
                    case_flag,
                } => {
                    if let Some(caps) = regex.captures(subject_dn) {
                        let mut out = String::new();
                        caps.expand(replacement, &mut out);
                        return match case_flag {
                            CaseFlag::None => out,
                            CaseFlag::Lower => out.to_lowercase(),
                            CaseFlag::Upper => out.to_uppercase(),
                        };
                    }
                }
            }
        }
        cn.to_owned()
    }
}

/// Split the rule spec at rule boundaries. Rules are separated by
/// commas, but commas can appear inside a `RULE:`'s regex body
/// (Subject DNs use them as a separator). Split only at commas
/// immediately followed by `RULE:` or `DEFAULT`. Mirrors the Go
/// `splitRules` heuristic.
fn split_rules(spec: &str) -> Vec<String> {
    let bytes = spec.as_bytes();
    let mut out = Vec::new();
    let mut i = 0;
    while i < bytes.len() {
        let mut j = i;
        while j < bytes.len() {
            if bytes[j] == b',' {
                let mut k = j + 1;
                while k < bytes.len() && (bytes[k] == b' ' || bytes[k] == b'\t') {
                    k += 1;
                }
                if spec[k..].starts_with("RULE:") || spec[k..].starts_with("DEFAULT") {
                    break;
                }
            }
            j += 1;
        }
        out.push(spec[i..j].trim().to_owned());
        if j < bytes.len() {
            j += 1;
        }
        i = j;
    }
    out
}

fn parse_one(raw: &str) -> Result<Rule, AuthError> {
    if raw == "DEFAULT" {
        return Ok(Rule::Default);
    }
    let body = raw.strip_prefix("RULE:").ok_or_else(|| {
        AuthError::PrincipalMappingParse(format!(
            "rule {raw:?}: must start with RULE: or be DEFAULT"
        ))
    })?;

    let (body, case_flag) = if let Some(b) = body.strip_suffix("/L") {
        (b, CaseFlag::Lower)
    } else if let Some(b) = body.strip_suffix("/U") {
        (b, CaseFlag::Upper)
    } else if let Some(b) = body.strip_suffix('/') {
        (b, CaseFlag::None)
    } else {
        return Err(AuthError::PrincipalMappingParse(format!(
            "rule {raw:?}: must end with '/', '/L', or '/U'"
        )));
    };

    let slash = body.find('/').ok_or_else(|| {
        AuthError::PrincipalMappingParse(format!(
            "rule {raw:?}: missing '/' between regex and replacement"
        ))
    })?;
    let pattern = &body[..slash];
    let replacement = &body[slash + 1..];
    let regex = Regex::new(pattern)?;
    Ok(Rule::Replace {
        regex,
        replacement: replacement.to_owned(),
        case_flag,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_spec_returns_cn() {
        let m = PrincipalMapper::parse("").unwrap();
        assert_eq!(m.apply("CN=alice,OU=Engineering", "alice"), "alice");
    }

    #[test]
    fn default_returns_cn_unchanged() {
        let m = PrincipalMapper::parse("DEFAULT").unwrap();
        assert_eq!(m.apply("CN=AlIcE,OU=X", "AlIcE"), "AlIcE");
    }

    #[test]
    fn first_rule_wins_with_back_reference() {
        let m = PrincipalMapper::parse("RULE:^CN=([^,]+),.*$/$1/").unwrap();
        assert_eq!(m.apply("CN=alice,OU=Engineering,O=Acme", "alice"), "alice");
    }

    #[test]
    fn lower_case_flag_applied() {
        let m = PrincipalMapper::parse("RULE:^CN=([^,]+).*$/$1/L").unwrap();
        assert_eq!(m.apply("CN=Alice,OU=X", "Alice"), "alice");
    }

    #[test]
    fn upper_case_flag_applied() {
        let m = PrincipalMapper::parse("RULE:^CN=([^,]+).*$/$1/U").unwrap();
        assert_eq!(m.apply("CN=alice,OU=X", "alice"), "ALICE");
    }

    #[test]
    fn unmatched_rule_falls_through_to_cn() {
        let m = PrincipalMapper::parse("RULE:^OU=Eng.*$/eng-user/").unwrap();
        assert_eq!(m.apply("CN=alice,OU=Sales", "alice"), "alice");
    }

    #[test]
    fn split_rules_keeps_dn_commas_inside_regex() {
        // The pattern body contains a comma — must not be split there.
        let m = PrincipalMapper::parse("RULE:^CN=([^,]+),OU=([^,]+),.*$/$1@$2/L,DEFAULT").unwrap();
        assert_eq!(m.apply("CN=Alice,OU=ENG,O=Acme", "Alice"), "alice@eng");
    }

    #[test]
    fn invalid_rule_syntax_errors() {
        let err = PrincipalMapper::parse("RULE:bad-syntax").unwrap_err();
        assert!(matches!(err, AuthError::PrincipalMappingParse(_)));
    }

    #[test]
    fn invalid_regex_errors() {
        let err = PrincipalMapper::parse("RULE:[unclosed/x/").unwrap_err();
        assert!(matches!(err, AuthError::Regex(_)));
    }

    #[test]
    fn multiple_rules_first_match_wins() {
        let m = PrincipalMapper::parse(
            "RULE:^CN=admin,.*$/superuser/U,RULE:^CN=([^,]+),.*$/$1/L,DEFAULT",
        )
        .unwrap();
        assert_eq!(m.apply("CN=admin,OU=X", "admin"), "SUPERUSER");
        assert_eq!(m.apply("CN=Bob,OU=X", "Bob"), "bob");
        // Neither rule's regex matches the DN below; the explicit
        // DEFAULT returns the CN unchanged.
        assert_eq!(m.apply("OU=NoCNHere", "fallback-cn"), "fallback-cn");
    }
}
